wallet, sqlite: Encapsulate SQLite statements in a RAII class

achow101 opened the pull request:
SQLite statements are C objects which can be long lived, so having them be initialized in a constructor, and cleanup everything in a destructor, makes the sqlite code cleaner and easier to read. This also lets us have the various functions needed to interact with sqlite statements be member functions.

This will be particularly useful in future work which makes use of more complicated SQL statements.

achow101 committed:
sqlite: Add SQLiteStatement RAII class

This class will be used to encapsulate a sqlite3_stmt

achow101 committed:
sqlite: Make Column template function

achow101 committed:
sqlite: Refactor ReadPragmaInteger to use SQLiteStatement

achow101 committed:
sqlite: Use SQLiteStatement in PRAGMA integrity_check

achow101 committed:
sqlite: Use SQLiteStatement in check_main_stmt

achow101 committed:
sqlite: Have SQLiteCursor store SQLiteStatement

achow101 committed:
sqlite: Replace remaining sqlite3_stmt usage with SQLiteStatement

achow101 committed:
sqlite: Refactor common writing code from WriteKey and ExecStatement

WriteKey and ExecStatement use the same code for the actual execution of
the statement. This is refactored into a separate function, also called
ExecStatement, and the original ExecStatement renamed to
ExecEraseStatement as it is only used by the erase functions.

achow101 committed:
sqlite: Inline BindBlobToStatement and SpanFromBlob

achow101 commented on src/wallet/sqlite.cpp:
```diff
@@ -58,44 +52,98 @@ static int TraceSqlCallback(unsigned code, void* context, void* param1, void* pa
+        } else {
+            // GCC <= 12 and Clang <= 16 don't like static_assert(false) in an else, so we use the following to achieve the same effect
+            // From: https://devblogs.microsoft.com/oldnewthing/20200311-00/?p=103553
+            static_assert(!sizeof(T*));
```
Done

w0xlt commented on src/util/transaction_identifier.h:
```diff
@@ -36,7 +36,10 @@ class transaction_identifier
+    using value_type = uint256::value_type;
+
     transaction_identifier() : m_wrapped{} {}
+    transaction_identifier(std::span<const value_type> sp) : m_wrapped(sp) {}
```
Is there any specific reason not to use `unsigned char` directly, like `std::span<const unsigned char > sp)` ?

achow101 replied:
In theory we could change `uint256`'s `value_type`, and we wouldn't need to make any changes here in that case.

achow101 replied:
I've dropped this commit from this PR and moved it to #33034. Also dropped `value_type` and made the constructor just take `span<const std::byte>` and that seems to work.

w0xlt reviewed (commented):
Looks good to me

ACK https://github.com/bitcoin/bitcoin/pull/33033/commits/b2c824a3ce0877f2cb7cbed1731088ed8ba6171a

rkrux reviewed (commented):
Started reviewing; I'm not opposed to this introduction at the moment and I will get more clarity soon in the review.

Initial Concept ACK 1829b3512d7bf430ec44daa801eb8fadf46cd30b

These 3 commits from #33034 looks pretty generic and related enough. Can they be included in this PR?
- 818989093d9e6b8ee765c21630f7342b99af7fa0 "sqlite: Make SQLiteStatement::Bind a template function"
- 07cac74e1340c7e771850633dd9418ff3efa5bb7 "sqlite: Make SQLiteDatabase::Column an std::optional"
- ff47eea527e58071bdaff5c2acbc98ee97294021 "sqlite: Add additional blob types to SQLiteStatement::Column"

vasild commented on src/wallet/sqlite.cpp:
```diff
@@ -58,44 +52,96 @@ static int TraceSqlCallback(unsigned code, void* context, void* param1, void* pa
+    {
+        int res = sqlite3_prepare_v2(m_db, stmt_text.c_str(), -1, &m_stmt, nullptr);
+        if (res != SQLITE_OK) {
+            throw std::runtime_error(strprintf(
```
Just an observation:

Some of the callers of `sqlite3_prepare_v2()` call `sqlite3_finalize()` on failure. The doc says that if prepare fails then [*ppStmt is set to NULL](https://sqlite.org/c3ref/prepare.html) and [Invoking sqlite3_finalize() on a NULL pointer is a harmless no-op](https://sqlite.org/c3ref/finalize.html).

So it is ok to omit the finalize call.

vasild commented on src/wallet/sqlite.cpp:
```diff
@@ -58,44 +52,96 @@ static int TraceSqlCallback(unsigned code, void* context, void* param1, void* pa
+        if constexpr (std::integral<T> && sizeof(T) <= 4) {
+            return sqlite3_column_int(m_stmt, col);
+        } else if constexpr (std::integral<T> && sizeof(T) <= 8) {
+            return static_cast<int64_t>(sqlite3_column_int64(m_stmt, col));
```
What happens if `T` is unsigned? Maybe add https://en.cppreference.com/w/cpp/types/is_signed.html here: `std::integral<T> && std::is_signed<T>`? (and leave it to drop to "else assert always false" for unsigned `T`s)

achow101 replied:
Done

vasild commented on src/wallet/sqlite.cpp:
```diff
@@ -432,12 +452,8 @@ void SQLiteBatch::Close()
-        *stmt_prepared = nullptr;
+        // Reset the unique_ptr to destroy the contained SQLiteStatement
+        stmt_prepared->reset();
     }
```
The second element in the pair (`stmt_description`) is not used. This snippet can be reduced to:

```cpp
    for (auto stmt : {&m_read_stmt, &m_insert_stmt, &m_overwrite_stmt, &m_delete_stmt, &m_delete_prefix_stmt}) {
        stmt->reset();
    }                   
```

Or simple stupid:
```cpp
    m_read_stmt.reset();
    m_insert_stmt.reset();
    m_overwrite_stmt.reset();
    m_delete_stmt.reset();
    m_delete_prefix_stmt.reset();
```

vasild reviewed (commented):
Approach ACK 1829b3512d7bf430ec44daa801eb8fadf46cd30b

vasild commented on src/wallet/sqlite.cpp:
```diff
@@ -58,44 +52,96 @@ static int TraceSqlCallback(unsigned code, void* context, void* param1, void* pa
+    ~SQLiteStatement()
+    {
+        Reset();
+        sqlite3_finalize(m_stmt);
```
Maybe (not tested):

```cpp
    int Step()
    {
        const int ret = sqlite3_step(m_stmt);
        if (ret == SQLITE_ROW || ret == SQLITE_DONE) {
            return ret;
        }
        sqlite3_finalize(m_stmt);
        m_stmt = nullptr;
        // then the destructor can assert(sqlite3_finalize(m_stmt) == SQLITE_OK);
    }
```
Related to https://github.com/bitcoin/bitcoin/pull/33033#discussion_r2518504627

achow101 replied:
> > If the most recent evaluation of statement S failed, then sqlite3_finalize(S) returns the appropriate [error code](https://sqlite.org/rescode.html) or [extended error code](https://sqlite.org/rescode.html#extrc)
> 
> I am not sure what is the best way to handle this. Ignoring the return value does not look good.

This text reads to me as `sqlite3_finalize` propagating any previous errors, e.g. from `sqlite3_step`. Since any errors from that should already be handled, I think it's ok to ignore the result of finalize. I think the previous code is actually incorrect since the error result is not actually a result of finalize failing.

achow101 commented on src/wallet/sqlite.h:
Done

achow101 commented on src/wallet/sqlite.cpp:
```diff
@@ -58,44 +52,96 @@ static int TraceSqlCallback(unsigned code, void* context, void* param1, void* pa
+        int res = sqlite3_prepare_v2(m_db, stmt_text.c_str(), -1, &m_stmt, nullptr);
+        if (res != SQLITE_OK) {
+            throw std::runtime_error(strprintf(
+                "SQLiteDatabase: Failed to prepare SQL statement: %s\n", sqlite3_errstr(res)));
```
Done

achow101 commented on src/wallet/sqlite.cpp:
```diff
@@ -58,44 +52,96 @@ static int TraceSqlCallback(unsigned code, void* context, void* param1, void* pa
+    {
+        Reset();
+        sqlite3_finalize(m_stmt);
+    }
```
Removed

vasild reviewed (approved):
ACK 0af7c2a11819a21e11abf34b4229c5c21150d827

davidgumberg commented on src/wallet/sqlite.cpp:
```diff
@@ -211,23 +211,11 @@ SQLiteDatabase::SQLiteDatabase(const fs::path& dir_path, const fs::path& file_pa
+    m_insert_stmt = std::make_unique<SQLiteStatement>(*m_database.m_db, "INSERT INTO main VALUES(?, ?)");
+    m_overwrite_stmt = std::make_unique<SQLiteStatement>(*m_database.m_db, "INSERT or REPLACE into main values(?, ?)");
+    m_delete_stmt = std::make_unique<SQLiteStatement>(*m_database.m_db, "DELETE FROM main WHERE key = ?");
+    m_delete_prefix_stmt = std::make_unique<SQLiteStatement>(*m_database.m_db, "DELETE FROM main WHERE instr(key, ?) = 1");
```
I wonder if the introduction of `SQLiteStatement` means we can do better than initializing every statement as a member of every batch: Given that we are `Reset()`'ing statements every single time they are used, and AFAICT batches tend to be short-lived, I think it would be cheaper and less complicated to initialize statement objects just before they are used in the various `Read*()` `Erase*()` , and it takes better advantage of the fact that statements are RAII than manually calling a closing function in Batch's destructor and asserting at the beginning of every `WalletBatch()` function that the relevant `stmt` object exists, here's one take of how something like this could be done:

https://github.com/davidgumberg/bitcoin/commit/b67e0301fd30564188470a29d3c10df2ff561e9c

achow101 committed:
sqlite: Construct SQLiteStatements when needed

Instead of constructing all SQLiteStatements when SQLiteBatch is
constructed, construct them once when they are needed before each read,
write, or erase operation. Once constructed, the statement will persist
for the lifetime of the SQLiteBatch to be reused across multiple
statements.

achow101 commented on src/wallet/sqlite.cpp:
```diff
@@ -524,53 +544,27 @@ bool SQLiteBatch::WriteKey(DataStream&& key, DataStream&& value, bool overwrite)
 bool SQLiteBatch::ErasePrefix(std::span<const std::byte> prefix)
 {
-    return ExecStatement(m_delete_prefix_stmt.get(), prefix);
+    return ExecEraseStatement(m_delete_prefix_stmt.get(), prefix);
```
Done

sedited commented:
Ping @davidgumberg @vasild for hopefully another round of review.

Eunovo commented on src/wallet/sqlite.cpp:
```diff
@@ -120,7 +94,18 @@ class SQLiteStatement
+        if (res != SQLITE_OK) {
+            LogWarning("Unable to bind %s to statement: %s", description, sqlite3_errstr(res));
+            Reset();
+            return false;
```
https://github.com/bitcoin/bitcoin/pull/33033/commits/b79895c87a3e50de09f7a61887f1389030a94823:

I wonder if `Bind` should throw an `Exception` instead? It doesn't look like we can recover from a failure to `sqlite3_bind_blob`.

achow101 replied:
Not all callers of this will have exception handling, but they all should have handling if it fails.

Eunovo replied:
Looks like a good candidate for `[[nodiscard]]`

achow101 replied:
Done

Eunovo commented on src/wallet/sqlite.cpp:
```diff
@@ -78,6 +78,51 @@ static bool BindBlobToStatement(sqlite3_stmt* stmt,
+    explicit SQLiteStatement(sqlite3& db, const std::string& stmt_text)
+        : m_db(db)
+    {
+        int res = sqlite3_prepare_v2(&m_db, stmt_text.c_str(), -1, &m_stmt, nullptr);
```
https://github.com/bitcoin/bitcoin/pull/33033/commits/bfa96f8c12ba40fbc66499fa72251bec9e3367e1:

I think it's useful to document that `SQLiteStatement` can only handle one statement in `stmt_text` or, preferably, provide the `pzTail` argument and throw an error if `pzTail` is not Null after `sqlite3_prepare_v2` (this indicates that there is uncompiled SQL code left).

achow101 replied:
If I need to retouch.

achow101 replied:
Done

Eunovo reviewed (commented):
https://github.com/bitcoin/bitcoin/pull/33033/commits/1682818a7b19c7c391fe4a75d96f620b1ac5127c:

Looks good to me. Left some comments.

pablomartin4btc commented on src/wallet/sqlite.h:
```diff
@@ -72,8 +71,18 @@ class SQLiteBatch : public DatabaseBatch
+
+    void MaybeSetupSQLStatement(SQLiteStatementType type);
+    bool ExecStatement(SQLiteStatement* stmt);
+    bool ExecEraseStatement(SQLiteStatement* stmt, std::span<const std::byte> blob);
```
nit: `ExecEraseStatement` is never used nor implemented.
```suggestion
```

achow101 replied:
Removed

pablomartin4btc commented on src/wallet/sqlite.cpp:
```diff
@@ -58,45 +52,105 @@ static int TraceSqlCallback(unsigned code, void* context, void* param1, void* pa
+            }
+            size_t size = sqlite3_column_bytes(m_stmt, col);
+            std::string str_text(text, size);
+            return str_text;
```
nit:
```suggestion
            return std::string(text, size);
```

achow101 replied:
Done

pablomartin4btc commented on src/wallet/sqlite.cpp:
```diff
@@ -58,45 +52,105 @@ static int TraceSqlCallback(unsigned code, void* context, void* param1, void* pa
+        }
+        int result = pragma_read_stmt.Column<int>(0);
+        return result;
+    } catch (std::runtime_error& e) {
```
nit (more consistently within the repo):
```suggestion
    } catch (const std::runtime_error& e) {
```

achow101 replied:
Done

pablomartin4btc commented on src/wallet/sqlite.cpp:
```diff
@@ -58,45 +52,105 @@ static int TraceSqlCallback(unsigned code, void* context, void* param1, void* pa
+            return std::nullopt;
+        }
+        int result = pragma_read_stmt.Column<int>(0);
+        return result;
```
nit:
```suggestion
        return pragma_read_stmt.Column<int>(0);
```

achow101 replied:
Done

pablomartin4btc commented on src/wallet/sqlite.cpp:
```diff
@@ -199,45 +253,37 @@ bool SQLiteDatabase::Verify(bilingual_str& error)
+            error += Untranslated(strprintf("%s\n", str_msg));
         }
-        error += Untranslated(strprintf("%s\n", str_msg));
+    } catch (std::runtime_error& e) {
```
nit: same as [here](https://github.com/bitcoin/bitcoin/pull/33033/changes#r3659824157).

pablomartin4btc commented on src/wallet/sqlite.cpp:
```diff
@@ -58,45 +52,105 @@ static int TraceSqlCallback(unsigned code, void* context, void* param1, void* pa
+        } else if constexpr (std::is_same_v<T, std::string>) {
+            const char* text = (const char*)sqlite3_column_text(m_stmt, col);
+            if (!text) {
+                return "";
```
worth considering a `LogWarning` or at least a doc comment?

achow101 replied:
No. A column that stores a string could contain nothing in it if there is no string to store, which is equivalent to the empty string.

pablomartin4btc commented on src/wallet/sqlite.h:
```diff
@@ -72,8 +71,18 @@ class SQLiteBatch : public DatabaseBatch
 
-    void SetupSQLStatements();
-    bool ExecStatement(sqlite3_stmt* stmt, std::span<const std::byte> blob);
+    enum SQLiteStatementType
```
nit: per [doc/developer-notes.md](https://github.com/bitcoin/bitcoin/blob/master/doc/developer-notes.md), prefer `enum class` over plain `enum`.
```suggestion
    enum class SQLiteStatementType
```

achow101 replied:
Done

pablomartin4btc reviewed (commented):
ACK 1682818a7b19

Left some inline nits. Two broader observations I think worth noting:
- `GetNewPrefixCursor` bind failure is now a throw instead of `nullptr` -
Previously `GetNewPrefixCursor` returned `nullptr` when `BindBlobToStatement` failed, and callers handled that. Now the `SQLiteCursor` constructor throws `std::runtime_error` if either bind call fails, and no caller catches it — so it propagates up the stack. Also, since this PR introduces the behaviour change, wouldn't be natural to clean up the now-dead null-checks in `walletdb.cpp` (lines 193, 470, 523)?

- `sqlite3_prepare_v2` ignores trailing SQL (`pzTail`) -
The `SQLiteStatement` constructor passes `nullptr` as the `pzTail` out-parameter, so if `stmt_text` ever contained more than one statement (e.g. a stray semicolon — `"INSERT INTO main VALUES(?,?); DROP TABLE main"`), SQLite would silently prepare only the first and discard the rest (@Eunovo [raised](https://github.com/bitcoin/bitcoin/pull/33033/changes#r3645464121) it also).

achow101 commented:
> * Also, since this PR introduces the behaviour change, wouldn't be natural to clean up the now-dead null-checks in `walletdb.cpp` (lines 193, 470, 523)?

`GetNewPrefixCursor` can still return `nullptr`.
